src/reactive-handler.ts

```diff
@@ -1,68 +1,62 @@
-}
-
-/**
- * Unwrap property descriptors will set value on original descriptor
- * We only need to unwrap if value is specified
- * @param descriptor external descrpitor provided to define new property on original value
- */
-function unwrapDescriptor(descriptor: PropertyDescriptor): PropertyDescriptor {
-    if (hasOwnProperty.call(descriptor, 'value')) {
-        descriptor.value = unwrap(descriptor.value);
-    }
-    return descriptor;
-}
-
-function lockShadowTarget(membrane: ReactiveMembrane, shadowTarget: ReactiveMembraneShadowTarget, originalTarget: any): void {
-    const targetKeys = ArrayConcat.call(getOwnPropertyNames(originalTarget), getOwnPropertySymbols(originalTarget));
-    targetKeys.forEach((key: PropertyKey) => {
-        let descriptor = getOwnPropertyDescriptor(originalTarget, key) as PropertyDescriptor;
-
-        // We do not need to wrap the descriptor if configurable
-        // Because we can deal with wrapping it when user goes through
-        // Get own property descriptor. There is also a chance that this descriptor
-        // could change sometime in the future, so we can defer wrapping
-        // until we need to
-        if (!descriptor.configurable) {
-            descriptor = wrapDescriptor(membrane, descriptor, wrapValue);
-        }
-        ObjectDefineProperty(shadowTarget, key, descriptor);
-    });
-
-    preventExtensions(shadowTarget);
-}
+const getterMap = new WeakMap<() => any, () => any>();
+const setterMap = new WeakMap<(v: any) => void, (v: any) => void>();
 
-export class ReactiveProxyHandler {
-    private originalTarget: any;
-    private membrane: ReactiveMembrane;
+export class ReactiveProxyHandler implements MembraneProxyHandler {
+    originalTarget: any;
+    membrane: ReactiveMembrane;
 
     constructor(membrane: ReactiveMembrane, value: any) {
         this.originalTarget = value;
         this.membrane = membrane;
     }
+    wrapValue(value: any): any {
+        const { membrane } = this;
+        return membrane.valueIsObservable(value) ? membrane.getProxy(value) : value;
+    }
+    wrapGetter(originalGet: () => any): () => any {
+        if (getterMap.has(originalGet)) {
+            return getterMap.get(originalGet) as () => any;
+        }
+        const handler = this;
+        const get = function (this: any): any {
+            // invoking the original getter with the original target
+            return handler.wrapValue(originalGet.call(unwrap(this)));
+        };
+        getterMap.set(originalGet, get);
+        return get;
+    }
+    wrapSetter(originalSet: (v: any) => void): (v: any) => void {
+        if (setterMap.has(originalSet)) {
+            return setterMap.get(originalSet) as (v: any) => void;
+        }
+        const handler = this;
+        const set = function (this: any, v: any) {
+            // invoking the original setter with the original target
+            handler.wrapValue(originalSet.call(unwrap(this), v));
```

Comment:
The value being set could be another proxy, I think it should be unwrapped here.
```suggestion
            handler.wrapValue(originalSet.call(unwrap(this), unwrap(v)));
```

Reply:
good catch! I will add a test for that as well.